File: app/src/main/java/nl/rsdt/japp/jotial/maps/wrapper/google/GoogleMarker.java
package nl.rsdt.japp.jotial.maps.wrapper.google;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;

import com.google.android.gms.maps.model.BitmapDescriptorFactory;
import com.google.android.gms.maps.model.LatLng;

import nl.rsdt.japp.application.Japp;
import nl.rsdt.japp.jotial.maps.wrapper.IJotiMap;
import nl.rsdt.japp.jotial.maps.wrapper.IMarker;

/**
 * Created by mattijn on 08/08/17.
 */

public class GoogleMarker implements IMarker {
    private static IJotiMap.OnMarkerClickListener allOnClickLister;
    private final com.google.android.gms.maps.model.Marker googleMarker;
    private IMarker.OnClickListener onClickListener;

    public GoogleMarker(com.google.android.gms.maps.model.Marker marker) {
        this.googleMarker = marker;
    }

    public static void setAllOnClickLister(IJotiMap.OnMarkerClickListener onClickListener){
        allOnClickLister = onClickListener;
    }

    private boolean onClick(){
        if (allOnClickLister != null){
            if (!allOnClickLister.OnClick(this)){
                if (this.onClickListener == null) {
                    showInfoWindow();
                    return false;
                }else{
                    return this.onClickListener.OnClick(this);
                }
            }
        }else {
            if (this.onClickListener == null) {
                showInfoWindow();
                return false;
            }else{
                return this.onClickListener.OnClick(this);
            }
        }
        return false;
    }

    @Override
    public void showInfoWindow(){
        googleMarker.showInfoWindow();
    }

    @Override
    public void remove() {
        this.googleMarker.remove();
    }

    @Override
    public String getTitle() {
        return this.googleMarker.getTitle();
    }

    @Override
    public LatLng getPosition() {
        return this.googleMarker.getPosition();
    }

    @Override
    public void setPosition(LatLng latLng) {
        this.googleMarker.setPosition(latLng);
    }

    @Override
    public void setOnClickListener(IMarker.OnClickListener onClickListener){
        this.onClickListener = onClickListener;
    }

    @Override
    public void setIcon(int drawableHunt) {
        this.setIcon(BitmapFactory.decodeResource(Japp.getAppResources(),drawableHunt));
    }

    @Override
    public void setIcon(Bitmap bitmap) {
        this.googleMarker.setIcon(BitmapDescriptorFactory.fromBitmap(bitmap));
    }

    @Override
    public void setTitle(String title) {
        googleMarker.setTitle(title);
    }

    @Override
    public boolean isVisible() {
        return this.googleMarker.isVisible();
    }

    @Override
    public void setVisible(boolean visible) {
        googleMarker.setVisible(visible);
    }

    @Override
    public void setRotation(float rotation) {
        googleMarker.setRotation(rotation);
    }

    @Override
    public String getId() {
        return googleMarker.getId();
    }
}

File: app/src/main/java/nl/rsdt/japp/jotial/maps/wrapper/google/GoogleMarker.kt
package nl.rsdt.japp.jotial.maps.wrapper.google

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.util.Log

import com.google.android.gms.maps.model.BitmapDescriptorFactory
import com.google.android.gms.maps.model.LatLng
import com.google.gson.Gson

import nl.rsdt.japp.application.Japp
import nl.rsdt.japp.jotial.maps.management.MarkerIdentifier
import nl.rsdt.japp.jotial.maps.wrapper.IJotiMap
import nl.rsdt.japp.jotial.maps.wrapper.IMarker
import org.acra.ktx.sendWithAcra

/**
 * Created by mattijn on 08/08/17.
 */

class GoogleMarker(private val googleMarker: com.google.android.gms.maps.model.Marker) : IMarker {
    override val identifier: MarkerIdentifier?
        get() {
            return if (title.isNotBlank()){
                try {
                    Gson().fromJson(title, MarkerIdentifier::class.java)?:null
                } catch (e: Exception) {
                    Log.e("GoogleMarker", e.toString())
                    e.sendWithAcra()
                    null
                }
            }
            else {
                null
            }
        }
    private var onClickListener: IMarker.OnClickListener? = null

    override var title: String
        get() = this.googleMarker.title
        set(title) {
            googleMarker.title = title
        }

    override var position: LatLng
        get() = this.googleMarker.position
        set(latLng) {
            this.googleMarker.position = latLng
        }

    override var isVisible: Boolean
        get() = this.googleMarker.isVisible
        set(visible) {
            googleMarker.isVisible = visible
        }

    override val id: String
        get() = googleMarker.id

    internal fun onClick(): Boolean {
        allOnClickLister?.OnClick(this)
        onClickListener?.OnClick(this)
        return false
    }

    override fun showInfoWindow() {
        googleMarker.showInfoWindow()
    }

    override fun remove() {
        this.googleMarker.remove()
    }

    override fun setOnClickListener(onClickListener: IMarker.OnClickListener?) {
        this.onClickListener = onClickListener
    }

    override fun setIcon(drawableHunt: Int) {
        this.setIcon(BitmapFactory.decodeResource(Japp.appResources, drawableHunt))
    }

    override fun setIcon(bitmap: Bitmap?) {
        this.googleMarker.setIcon(BitmapDescriptorFactory.fromBitmap(bitmap))
    }

    override fun setRotation(rotation: Float) {
        googleMarker.rotation = rotation
    }

    companion object {
        private var allOnClickLister: IJotiMap.OnMarkerClickListener? = null

        fun setAllOnClickLister(onClickListener: IJotiMap.OnMarkerClickListener?) {
            allOnClickLister = onClickListener
        }
    }
}
